### PR TITLE
Cleanup interfaces around ExtensionManager

### PR DESCRIPTION
**Pull request by peternied**
### Description
Cleanup interfaces around ExtensionManager
    
ExtensionsManager has a whole lot of functionality that shouldn't be
exposed to external classes, marked much of this package protected for a
cleaner interface.  Then updated the Noop version to override all public
functions to remove extranious feature flag checks.
    
Note; I didn't go so far to remove all getters/setters that were unused
but that could be a good follow up in the future.

### Check List
- [ ] ~New functionality includes testing.~
  - [ ] ~All tests pass~
- [ ] ~New functionality has been documented.~
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peternied** [comment]
Hi @dbwiddis @varuntumbe @owaiskazi19 @ryanbogan @mloufra @andrross - you've made changes to the ExtensionManager in the past and your review would be valuable as you have context around the construction/design choices.

File: server/src/test/java/org/opensearch/snapshots/SnapshotResiliencyTests.java
Comment (peternied) [inline server/src/test/java/org/opensearch/snapshots/SnapshotResiliencyTests.java:1807]: I was pretty baffled when I saw this feature flag being checked inside of test code - I'm assuming this was a mistake but would love more detail.

File: server/src/main/java/org/opensearch/node/Node.java
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/node/Node.java:878]: Why are we removing the feature flag from here?

Comment (peternied) [reply server/src/main/java/org/opensearch/node/Node.java:878]: The flag is redundant with the NoopExtensionManager

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/node/Node.java:878]: Not sure why `NoopExtensionManager` was added in the first place. Can we remove `NoopExtensionManager` completely and add the flag back? WDYT?

Comment (peternied) [reply server/src/main/java/org/opensearch/node/Node.java:878]: Feature flags are structural artifacts that make OpenSearch harder test and troubleshot by introducing duplicate code paths - while needed, the less branching codepaths the easier to maintain.

By depending on feature flags classes such as Node or IndicesService are responsible for checking is extensions enabled, when for the most part the answer is those classes shouldn't need to know if that is the case (there is the clear exception when the ExtensionManager is initialized).

Noop implementation invert the responsibility of the feature flag, so only the single implementing class needs to know rather than the callers of this class.

I am strongly in the noop implementation camp.

Comment (ryanbogan) [reply server/src/main/java/org/opensearch/node/Node.java:878]: I'm okay with removing the feature flag.  When I was adding all the checks, I erred on the side of caution and had them for every time that extensions interacted with code outside the directory.

File: server/src/test/java/org/opensearch/snapshots/SnapshotResiliencyTests.java
Comment (ryanbogan) [reply server/src/test/java/org/opensearch/snapshots/SnapshotResiliencyTests.java:1807]: I just added them everywhere that extensions code interacted with code outside the extensions directory.

Comment (ryanbogan) [reply server/src/test/java/org/opensearch/snapshots/SnapshotResiliencyTests.java:1807]: I also didn't want to change the initial method to add an extensionsManager parameter, so I overloaded the constructor.

**Review by ryanbogan — APPROVED** [review]
Looks good to me!  I didn't know that Noop implementations could simplify the code so much.

File: CHANGELOG.md
Comment (owaiskazi19) [inline CHANGELOG.md:97]: Nit: We can ignore the changelog for these changes and add skip changelog label.

**Review by owaiskazi19 — APPROVED** [review]
Thanks @peternied for letting us know a better alternative to handle feature flags.

**Comment by peternied** [comment]
It makes me so happy clean merges don't dismiss approvals on PRs, not sure when this was added by GitHub but its great!

**Comment by peternied** [comment]
New flaky test
```
* What went wrong:
Execution failed for task ':modules:analysis-common:test'.
> Process 'Gradle Test Executor 587' finished with non-zero exit value 1
  This problem might be caused by incorrect test process configuration.
  Please refer to the test execution section in the User Manual at https://docs.gradle.org/8.1.1/userguide/java_testing.html#sec:test_execution
```

```
org.gradle.internal.remote.internal.ConnectException: Could not connect to server [304ccce8-64cf-4a10-93d3-4a8ccc2aafc7 port:45325, addresses:[/127.0.0.1]]. Tried addresses: [/127.0.0.1].
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:67)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedClient.getConnection(MessageHubBackedClient.java:36)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:103)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.net.ConnectException: Connection refused
	at java.****/sun.nio.ch.Net.pollConnect(Native Method)
	at java.****/sun.nio.ch.Net.pollConnectNow(Net.java:672)
	at java.****/sun.nio.ch.SocketChannelImpl.finishTimedConnect(SocketChannelImpl.java:1141)
	at java.****/sun.nio.ch.SocketChannelImpl.blockingConnect(SocketChannelImpl.java:1183)
	at java.****/sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:98)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.tryConnect(TcpOutgoingConnector.java:81)
	at org.gradle.internal.remote.internal.inet.TcpOutgoingConnector.connect(TcpOutgoingConnector.java:54)
	... 5 more
```

**Comment by peternied** [comment]
> ERROR: script returned exit code 128

FYI @mch2 I fix the merge conflict on the CHANGELOG.md and then saw that the CI checks were still all passing - assumed this was OK and hit merge. After the change was merged I saw that the CI checks on the new commit hadn't started up yet.  I assume because this PR was merged it generated some strange state in the Jenkins workflow.

IMO I don't think anything will go awry because of this change but just in case this would be the place to start looking if something goes wrong.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:130]: Is this data structure and the `extensionIdMap` identical?

Comment (peternied) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:130]: extensionIdMap contains the list of configured extensions, whereas initializedExtensions is from discovered extensions that have responded as online.

**Comment by cwperks** [comment]
@saratvemulapalli @owaiskazi19 @ryanbogan Can you add a backport 2.x label to this PR?

**Comment by peternied** [comment]
I'll manually backport

**Pull request merged**
